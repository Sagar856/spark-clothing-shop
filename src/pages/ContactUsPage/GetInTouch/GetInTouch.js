import React, { useState } from "react";
// imported useFormik from formik package for the form
import { useFormik } from "formik";
import "../GetInTouch/GetInTouch.scss";

const GetInTouch = () => {
  // Setting state using setter and getter Of useSate() hook
  const [isSaved, setIsSaved] = useState(false);
  const [isError, setIsError] = useState(false);

  //useFormik for submitting the form data to the local backend
  const formik = useFormik({
    initialValues: {
      name: "",
      email: "",
      phone: "",
      message: "",
    },

    onSubmit: (values, { resetForm }) => {
      // fetch API using post method to post the data
      // posting form data to API
      fetch("http://localhost:5000/getInTouchData", {
        method: "POST",
        body: JSON.stringify(values),
        headers: {
          "Content-type": "application/json; charset=UTF-8",
        },
      })
        .then((res) => {
          setIsSaved(true);
          // console.log(res);
        })
        // fetching error if any during loading data
        .catch((err) => {
          setIsError(true);
        });
      // reseting the form after submit
      resetForm({ values: "" });
    },
  });

  // Error message display
  if (isError) {
    return (
      <div className="alert alert-danger" data-testid="errorMessage">
        Some Error Occurred! Try After Some Time
      </div>
    );
  }

  return (
    <div className="get-in-touch-form">
      <div className="feedback-form mt-5">
        {/*get in touch form */}
        <form
          autoComplete="off"
          onSubmit={formik.handleSubmit}
          className="contact-form"
        >
          <div className="form-data field-group">
            <label htmlFor="name">Name:</label>
            <input
              type="text"
              name="name"
              className="form-control"
              id="name"
              required
              placeholder="Enter name"
              data-testid="nameInput"
              value={formik.values.name}
              onChange={formik.handleChange}
            />
          </div>
          <div className="field-group">
            <label htmlFor="email">E-mail:</label>
            <input
              type="email"
              className="form-control"
              name="email"
              id="email"
              placeholder="Enter mail"
              required
              data-testid="emailInput"
              value={formik.values.email}
              onChange={formik.handleChange}
            />
          </div>
          <div className="field-group">
            <label htmlFor="phone">Phone:</label>
            <input
              type="number"
              className="form-control"
              name="phone"
              id="phone"
              required
              placeholder="Enter phone number"
              data-testid="phoneInput"
              value={formik.values.phone}
              onChange={formik.handleChange}
            />
          </div>
          <div className="field-group">
            <label htmlFor="message">Message:</label>
            <textarea
              type="text"
              name="message"
              className="form-control"
              id="message"
              required
              placeholder="Enter message"
              data-testid="messageInput"
              value={formik.values.message}
              onChange={formik.handleChange}
            />
          </div>
          <div>
            <button
              className="btn-style btn btn-info px-4 py-2 fw-bold"
              type="submit"
              data-testid="submitBtn"
              // disabling the submit button if email is not entered
              disabled={formik.values.email == ""}
            >
              Submit
            </button>
            <div>
              {/* Giving successful message after successful submission of data to backend */}
              {isSaved ? (
                <div data-testid="isSaved" className="alert alert-success">
                  Submitted Successfully...
                </div>
              ) : (
                ""
              )}
            </div>
          </div>
        </form>
      </div>
      <div className="get-in-touch-img-form">
        <img
          src="/assets/images/GetInTouchImgForm.png"
          className="d-block mt-5 get-in-touch-img"
          alt="getintouch"
          data-testid="getInTouchImgForm"
        ></img>
      </div>
    </div>
  );
};

export default GetInTouch;
