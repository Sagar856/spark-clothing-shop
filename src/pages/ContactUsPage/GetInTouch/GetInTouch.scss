.feedback-form {
  border: solid 1px #333;
  margin: auto;
  padding: 36px;
  background-color: #969797;
  border-radius: 20px;
}

.field-group {
  padding: 10px;
  display: flex;
}

.feedback-form .field-group label {
  width: 140px;
  display: inline-block;
}

.feedback-form .field-group input {
  width: 315px;
  border: 1px solid #0043c6;
  border-radius: 6px;
  padding: 5px 14px;
}

.feedback-form {
  vertical-align: top;
  margin-top: 12px;
}

.btn-style {
  color: black;
  border-radius: 7px;
  padding: 4px 10px;
}

.get-in-touch-form {
  display: grid;
  grid-template-columns: auto auto;
}

.get-in-touch-img-form {
  width: 500px;
  margin: auto;
}

.get-in-touch-img {
  width: 500px;
  height: 368px;
  border-radius: 20px;
}

.feedback-form .field-group textarea {
  width: 315px;
  border: 1px solid #0043c6;
  border-radius: 6px;
  padding: 5px 14px;
}
