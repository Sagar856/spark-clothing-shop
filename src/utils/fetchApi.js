export const fetchApi = (url, method) => {
  return fetch(url)
    .then((res) => {
      // if successful

      if (res.status == 200 || res.status == 201) {
        return res.json();
      } else {
        return {
          status: res.status,
          error: new Error("Invalid Response"),
        };
      }
    })
    .catch((err) => {
      // if error occurs
      return err;
    })
    .finally(() => {
    });
};
