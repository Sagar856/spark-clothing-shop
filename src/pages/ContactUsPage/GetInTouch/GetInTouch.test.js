import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";
import { fetchApi } from "../../../utils/fetchApi";
import GetInTouch from "./GetInTouch";

// setting up mock for fetchApi
// mocking a module with an automocked version when needed
jest.mock("../../../utils/fetchApi");

describe("Get in touch", () => {
  afterEach(jest.resetAllMocks);

  it("has proper get in touch form", () => {
    // Rendering the component
    render(<GetInTouch />);
    const name = screen.getByLabelText("Name:");
    const phone = screen.getByLabelText("Phone:");
    const email = screen.getByLabelText("E-mail:");
    const message = screen.getByLabelText("Message:");
    const submitBtn = screen.getByRole("button");
    // Assert
    expect(name).toBeInTheDocument();
    expect(phone).toBeInTheDocument();
    expect(email).toBeInTheDocument();
    expect(message).toBeInTheDocument();
    expect(submitBtn).toBeInTheDocument();
    expect(name).toHaveAttribute("type", "text");
    expect(phone).toHaveAttribute("type", "number");
    expect(email).toHaveAttribute("type", "email");
    expect(message).toHaveAttribute("type", "text");
  });

  it("has the submit button in disabled state when name and email field is empty", () => {
    // Rendering the component
    render(<GetInTouch />);
    const name = screen.getByLabelText("Name:");
    const email = screen.getByLabelText("E-mail:");
    const phone = screen.getByLabelText("Phone:");
    const message = screen.getByLabelText("Message:");
    // fetching mock Api
    const mockEventObj = {
      target: {
        value: "",
      },
    };
    // trigger change event
    fireEvent.change(name, mockEventObj);
    fireEvent.change(email, mockEventObj);
    fireEvent.change(phone, mockEventObj);
    fireEvent.change(message, mockEventObj);
    // Assert
    const button = screen.getByRole("button");
    expect(button).toHaveAttribute("disabled");
  });

  it("has success message on button click", async () => {
    // Rendering the component
    render(<GetInTouch />);
    const emailInput = screen.getByTestId("emailInput");
    const nameInput = screen.getByTestId("nameInput");
    const phoneInput = screen.getByTestId("phoneInput");
    const messageInput = screen.getByTestId("messageInput");
    // fetching mock Api
    const mockEventObjEmail = {
      target: {
        value: "virat@gmail.com",
      },
    };
    const mockEventObjName = {
      target: {
        value: "Sagar",
      },
    };
    const mockEventObjPhone = {
      target: {
        value: "868767857557",
      },
    };
    const mockEventObjMessage = {
      target: {
        value: "skldsldknlsaknlsakvcl",
      },
    };
    fetchApi.mockResolvedValue(mockEventObjEmail);
    fetchApi.mockResolvedValue(mockEventObjName);
    fetchApi.mockResolvedValue(mockEventObjPhone);
    fetchApi.mockResolvedValue(mockEventObjMessage);

    // Assert
    fireEvent.change(emailInput, mockEventObjEmail);
    fireEvent.change(nameInput, mockEventObjName);
    fireEvent.change(phoneInput, mockEventObjPhone);
    fireEvent.change(messageInput, mockEventObjMessage);
    userEvent.click(screen.getByTestId("submitBtn"));
    await waitFor(() => {
      const isSaved = screen.getByTestId("isSaved");
      expect(isSaved).toBeInTheDocument("Submitted Successfully...");
    });
  });

  // Snapshot testing for GetInTouch
  it("has right snapshot with all requirements completed", () => {
    // Taking snapshot and also converting into JSON
    const snapshotInJson = renderer.create(<GetInTouch />).toJSON();
    // Assert with a matcher toMatchSnapshot()
    expect(snapshotInJson).toMatchSnapshot();
  });

  // NEGATIVE SPEC
  it("has Error message when data not submitted to backend", async () => {
    // Rendering the component
    render(<GetInTouch />);
    const emailInput = screen.getByTestId("emailInput");
    const nameInput = screen.getByTestId("nameInput");
    const phoneInput = screen.getByTestId("phoneInput");
    const messageInput = screen.getByTestId("messageInput");
    // fetching mock Api
    const mockEventObjEmail = {
      target: {
        value: "virat@gmail.com",
      },
    };
    const mockEventObjName = {
      target: {
        value: "Sagar",
      },
    };
    const mockEventObjPhone = {
      target: {
        value: "34535345334",
      },
    };
    const mockEventObjMessage = {
      target: {
        value: "skldsldknlsaknlsakvcl",
      },
    };
    const error = {
      errorInfo: "Some Error Occurred! Try After Some Time",
    };
    fetchApi.mockRejectedValue(error);
    fetchApi.mockRejectedValue(mockEventObjEmail);
    fetchApi.mockRejectedValue(mockEventObjName);
    fetchApi.mockRejectedValue(mockEventObjPhone);
    fetchApi.mockRejectedValue(mockEventObjMessage);

    // Assert
    fireEvent.change(emailInput, mockEventObjEmail);
    fireEvent.change(nameInput, mockEventObjName);
    fireEvent.change(phoneInput, mockEventObjPhone);
    fireEvent.change(messageInput, mockEventObjMessage);
    userEvent.click(screen.getByTestId("submitBtn"));
    await waitFor(() => {
      //   expect(
      //     screen.getByText(/Some Error Occurred! Try After Some Time/i)
      //   ).toBeInTheDocument();
    });
  });
});
