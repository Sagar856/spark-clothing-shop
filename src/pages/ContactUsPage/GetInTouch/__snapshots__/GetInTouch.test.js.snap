// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get in touch has right snapshot with all requirements completed 1`] = `
<div
  className="get-in-touch-form"
>
  <div
    className="feedback-form mt-5"
  >
    <form
      autoComplete="off"
      className="contact-form"
      onSubmit={[Function]}
    >
      <div
        className="form-data field-group"
      >
        <label
          htmlFor="name"
        >
          Name:
        </label>
        <input
          className="form-control"
          data-testid="nameInput"
          id="name"
          name="name"
          onChange={[Function]}
          placeholder="Enter name"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="field-group"
      >
        <label
          htmlFor="email"
        >
          E-mail:
        </label>
        <input
          className="form-control"
          data-testid="emailInput"
          id="email"
          name="email"
          onChange={[Function]}
          placeholder="Enter mail"
          required={true}
          type="email"
          value=""
        />
      </div>
      <div
        className="field-group"
      >
        <label
          htmlFor="phone"
        >
          Phone:
        </label>
        <input
          className="form-control"
          data-testid="phoneInput"
          id="phone"
          name="phone"
          onChange={[Function]}
          placeholder="Enter phone number"
          required={true}
          type="number"
          value=""
        />
      </div>
      <div
        className="field-group"
      >
        <label
          htmlFor="message"
        >
          Message:
        </label>
        <textarea
          className="form-control"
          data-testid="messageInput"
          id="message"
          name="message"
          onChange={[Function]}
          placeholder="Enter message"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div>
        <button
          className="btn-style btn btn-info px-4 py-2 fw-bold"
          data-testid="submitBtn"
          disabled={true}
          type="submit"
        >
          Submit
        </button>
        <div />
      </div>
    </form>
  </div>
  <div
    className="get-in-touch-img-form"
  >
    <img
      alt="getintouch"
      className="d-block mt-5 get-in-touch-img"
      data-testid="getInTouchImgForm"
      src="/assets/images/GetInTouchImgForm.png"
    />
  </div>
</div>
`;
